### PR TITLE
[BEAM-10656] Enable bundle finalization within the Java direct runner.

### PR DESCRIPTION
**Pull request by lukecwik**
This is towards making all UnboundedSources execute as splittable dofns within the direct runner using the SDF unbounded source wrapper since it relies on bundle finalization to handle checkpoints.

------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [ ] [**Choose reviewer(s)**](https://beam.apache.org/contribute/#make-your-change) and mention them in a comment (`R: @username`).
 - [ ] Format the pull request title like `[BEAM-XXX] Fixes bug in ApproximateQuantiles`, where you replace `BEAM-XXX` with the appropriate JIRA issue, if applicable. This will automatically link the pull request to the issue.
 - [ ] Update `CHANGES.md` with noteworthy changes.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

See the [Contributor Guide](https://beam.apache.org/contribute) for more tips on [how to make review process smoother](https://beam.apache.org/contribute/#make-reviewers-job-easier).

Post-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

Lang | SDK | Dataflow | Flink | Samza | Spark | Twister2
--- | --- | --- | --- | --- | --- | ---
Go | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Go/lastCompletedBuild/) | --- | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Go_VR_Flink/lastCompletedBuild/) | --- | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Go_VR_Spark/lastCompletedBuild/) | ---
Java | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow_Java11/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink_Java11/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_PVR_Flink_Batch/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_PVR_Flink_Streaming/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Samza/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Spark/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_PVR_Spark_Batch/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_SparkStructuredStreaming/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Twister2/lastCompletedBuild/)
Python | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Python2/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Python35/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Python36/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Python37/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Python38/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Py_VR_Dataflow/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Py_VR_Dataflow_V2/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Py_ValCont/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Python2_PVR_Flink_Cron/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Python35_VR_Flink/lastCompletedBuild/) | --- | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Python_VR_Spark/lastCompletedBuild/) | ---
XLang | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_XVR_Direct/lastCompletedBuild/) | --- | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_XVR_Flink/lastCompletedBuild/) | --- | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_XVR_Spark/lastCompletedBuild/) | ---

Pre-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

--- |Java | Python | Go | Website
--- | --- | --- | --- | ---
Non-portable | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Java_Cron/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Python_Cron/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PreCommit_PythonLint_Cron/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PreCommit_PythonDocker_Cron/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Go_Cron/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Website_Cron/lastCompletedBuild/)
Portable | --- | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Portable_Python_Cron/lastCompletedBuild/) | --- | ---

See [.test-infra/jenkins/README](https://github.com/apache/beam/blob/master/.test-infra/jenkins/README.md) for trigger phrase, status and link of all Jenkins jobs.

GitHub Actions Tests Status (on master branch)
------------------------------------------------------------------------------------------------
Build python source distribution and wheels

See [CI.md](https://github.com/apache/beam/blob/master/CI.md) for more information about GitHub Actions CI.

### CODE REVIEW

**Comment by lukecwik** [comment]
R: @amaliujia @ibzib 
CC: @boyuanzz This should enable testing unbounded SDFs using the direct runner with bundle finalization callbacks.

**Comment by lukecwik** [comment]
retest this please

**Comment by lukecwik** [comment]
Run Java Precommit

**Comment by lukecwik** [comment]
Run Java PreCommit

**Comment by lukecwik** [comment]
retest this please

**Comment by lukecwik** [comment]
Run Spark ValidatesRunner

**Comment by lukecwik** [comment]
Run Flink ValidatesRunner

**Comment by lukecwik** [comment]
Run Spark ValidatesRunner

**Comment by lukecwik** [comment]
Run Flink ValidatesRunner

**Comment by lukecwik** [comment]
Run Spark ValidatesRunner

**Comment by lukecwik** [comment]
Run Flink ValidatesRunner

**Comment by lukecwik** [comment]
Run Spark ValidatesRunner

**Comment by lukecwik** [comment]
The test failure is known issue, failing in postcommit https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Spark/7773/testReport/ and https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink/7805/

File: runners/core-java/src/test/java/org/apache/beam/runners/core/OutputAndTimeBoundedSplittableProcessElementInvokerTest.java [hunk 1]
```diff
@@ -126,23 +128,30 @@ public void outputWindowedValue(
...
-          }
-        });
+              @Override
+              public Void getState() {
+                return null;
+              }
+            });
+    for (InMemoryBundleFinalizer.Finalization finalization :
+        bundleFinalizer.getAndClearFinalizations()) {
+      finalization.getCallback().onBundleSuccess();
```
Comment (ibzib) [inline runners/core-java/src/test/java/org/apache/beam/runners/core/OutputAndTimeBoundedSplittableProcessElementInvokerTest.java:152]: Why do we need to trigger the callbacks here?

File: runners/core-java/src/test/java/org/apache/beam/runners/core/OutputAndTimeBoundedSplittableProcessElementInvokerTest.java [hunk 2]
```diff
@@ -126,23 +128,30 @@ public void outputWindowedValue(
...
-          public Void getState() {
-            return null;
-          }
-        });
+              @Override
+              public Void getState() {
+                return null;
+              }
+            });
+    for (InMemoryBundleFinalizer.Finalization finalization :
```
Comment (ibzib) [inline runners/core-java/src/test/java/org/apache/beam/runners/core/OutputAndTimeBoundedSplittableProcessElementInvokerTest.java:150]: Can we make any assertions about the finalizations? ie Can we expect that they're not empty?

File: runners/direct-java/src/main/java/org/apache/beam/runners/direct/DirectExecutionContext.java [hunk 3]
```diff
@@ -81,6 +89,22 @@ public DirectTimerInternals timerInternals() {
       return timerInternals;
     }
 
+    @Override
+    public BundleFinalizer bundleFinalizer() {
+      if (bundleFinalizer == null) {
+        bundleFinalizer = new InMemoryBundleFinalizer();
```
Comment (ibzib) [inline runners/direct-java/src/main/java/org/apache/beam/runners/direct/DirectExecutionContext.java:95]: Why lazy-initialize? Is this just to save memory?

**Comment by lukecwik** [comment]
Expanded existing JIRA: https://issues.apache.org/jira/browse/BEAM-8460, opened #12503 to disable failing test category for spark and flink

File: runners/direct-java/src/main/java/org/apache/beam/runners/direct/EvaluationContext.java [hunk 4]
```diff
@@ -183,6 +188,15 @@ public void initialize(
         committedResult.getUnprocessedInputs().orElse(null),
         committedResult.getOutputs(),
         result.getWatermarkHold());
+
+    // Callback and requested bundle finalizations
+    for (InMemoryBundleFinalizer.Finalization finalization : result.getBundleFinalizations()) {
+      try {
+        finalization.getCallback().onBundleSuccess();
+      } catch (Exception e) {
+        LOG.warn("Failed to finalize requested bundle {}", finalization, e);
```
Comment (ibzib) [inline runners/direct-java/src/main/java/org/apache/beam/runners/direct/EvaluationContext.java:197]: I don't think logging only the finalization tells us enough. Can we log more information about the bundle itself?

**Comment by amaliujia** [comment]
Thanks @lukecwik 

Will take a look at this PR soon.

File: sdks/java/core/src/test/java/org/apache/beam/sdk/transforms/ParDoTest.java [hunk 5]
```diff
@@ -1499,6 +1504,146 @@ public void populateDisplayData(Builder builder) {
...
+        testFinalization(bundleFinalizer, output);
+      }
+    }
+
+    private static class BundleFinalizerOutputChecker
+        implements SerializableFunction<Iterable<String>, Void> {
+      @Override
+      public Void apply(Iterable<String> input) {
+        assertTrue(
+            "Expected to have received one callback enabling output to be produced but received none.",
```
Comment (ibzib) [inline sdks/java/core/src/test/java/org/apache/beam/sdk/transforms/ParDoTest.java:1543]: Multiple callbacks are allowed, right? (If so, please reword this as "at least one callback.")

Comment (ibzib) [reply sdks/java/core/src/test/java/org/apache/beam/sdk/transforms/ParDoTest.java:1543]: In other words, I'd like to clarify the assertion we're making about `input` and what it means.

File: runners/direct-java/src/main/java/org/apache/beam/runners/direct/DirectExecutionContext.java (hunk 3 above)
Comment (lukecwik) [reply runners/direct-java/src/main/java/org/apache/beam/runners/direct/DirectExecutionContext.java:95]: yup since most transforms won't use the bundle finalizer

File: sdks/java/core/src/test/java/org/apache/beam/sdk/transforms/ParDoTest.java (hunk 5 above)
Comment (lukecwik) [reply sdks/java/core/src/test/java/org/apache/beam/sdk/transforms/ParDoTest.java:1543]: Done

File: runners/core-java/src/test/java/org/apache/beam/runners/core/OutputAndTimeBoundedSplittableProcessElementInvokerTest.java (hunk 1 above)
Comment (lukecwik) [reply runners/core-java/src/test/java/org/apache/beam/runners/core/OutputAndTimeBoundedSplittableProcessElementInvokerTest.java:152]: We don't. I originally thought we did for some reason since I was changing this code together. I removed it and instead throw if it is ever accessed.

File: runners/core-java/src/test/java/org/apache/beam/runners/core/OutputAndTimeBoundedSplittableProcessElementInvokerTest.java (hunk 2 above)
Comment (lukecwik) [reply runners/core-java/src/test/java/org/apache/beam/runners/core/OutputAndTimeBoundedSplittableProcessElementInvokerTest.java:150]: Not relevant anymore.

**Comment by lukecwik** [comment]
Run Flink ValidatesRunner

**Comment by lukecwik** [comment]
Run Spark ValidatesRunner

File: runners/direct-java/src/main/java/org/apache/beam/runners/direct/EvaluationContext.java (hunk 4 above)
Comment (lukecwik) [reply runners/direct-java/src/main/java/org/apache/beam/runners/direct/EvaluationContext.java:197]: Added the completed bundle to the output.

**Comment by lukecwik** [comment]
Run Flink ValidatesRunner

**Comment by lukecwik** [comment]
Run Spark ValidatesRunner

File: runners/direct-java/src/main/java/org/apache/beam/runners/direct/DirectExecutionContext.java (hunk 3 above)
Comment (amaliujia) [reply runners/direct-java/src/main/java/org/apache/beam/runners/direct/DirectExecutionContext.java:95]: Line#37 says the implementation is not thread-safe. So I assume there is no need to consider concurrent access to this initialization, right?

File: runners/direct-java/src/main/java/org/apache/beam/runners/direct/EvaluationContext.java [hunk 6]
```diff
@@ -183,6 +188,15 @@ public void initialize(
         committedResult.getUnprocessedInputs().orElse(null),
         committedResult.getOutputs(),
         result.getWatermarkHold());
+
+    // Callback and requested bundle finalizations
+    for (InMemoryBundleFinalizer.Finalization finalization : result.getBundleFinalizations()) {
+      try {
+        finalization.getCallback().onBundleSuccess();
+      } catch (Exception e) {
```
Comment (amaliujia) [inline runners/direct-java/src/main/java/org/apache/beam/runners/direct/EvaluationContext.java:196]: Is is possible to make this Exception more specific? Or that does not add more values?

File: sdks/java/core/src/test/java/org/apache/beam/sdk/transforms/ParDoTest.java [hunk 7]
```diff
@@ -1499,6 +1504,146 @@ public void populateDisplayData(Builder builder) {
...
+          throws Exception {
+        if (WAS_FINALIZED.computeIfAbsent(uuid, (unused) -> new AtomicBoolean()).get()) {
+          output.output("bundle was finalized");
+          return;
+        }
+        bundleFinalizer.afterBundleCommit(
+            Instant.now().plus(Duration.standardSeconds(MAX_ATTEMPTS)),
+            () -> WAS_FINALIZED.computeIfAbsent(uuid, (unused) -> new AtomicBoolean()).set(true));
+        // We sleep here to give time for the runner to perform any prior callbacks.
+        sleep(100L);
```
Comment (amaliujia) [inline sdks/java/core/src/test/java/org/apache/beam/sdk/transforms/ParDoTest.java:1526]: nit: I am not sure if there are other ways, but such `sleep` might be a source of flakiness. E.g. prior callbacks not finish after 100L.

File: sdks/java/core/src/test/java/org/apache/beam/sdk/transforms/SplittableDoFnTest.java [hunk 8]
```diff
@@ -839,19 +845,20 @@ public ProcessContinuation process(
...
       }
       if (tracker.tryClaim(tracker.currentRestriction().getFrom() + 1)) {
         bundleFinalizer.afterBundleCommit(
             Instant.now().plus(Duration.standardSeconds(MAX_ATTEMPTS)),
-            () -> wasFinalized.set(true));
+            () -> WAS_FINALIZED.computeIfAbsent(uuid, (unused) -> new AtomicBoolean()).set(true));
         // We sleep here instead of setting a resume time since the resume time doesn't need to
         // be honored.
-        sleep(1000L); // 1 second
+        sleep(100L);
```
Comment (amaliujia) [inline sdks/java/core/src/test/java/org/apache/beam/sdk/transforms/SplittableDoFnTest.java:861]: nit: same here.

**Comment by lukecwik** [comment]
retest this please

File: runners/direct-java/src/main/java/org/apache/beam/runners/direct/EvaluationContext.java (hunk 6 above)
Comment (lukecwik) [reply runners/direct-java/src/main/java/org/apache/beam/runners/direct/EvaluationContext.java:196]: This exception is to catch user thrown exceptions since the bundle finalizing callback is user written code this is why its so wide and it can't be narrowed.

File: sdks/java/core/src/test/java/org/apache/beam/sdk/transforms/ParDoTest.java (hunk 7 above)
Comment (lukecwik) [reply sdks/java/core/src/test/java/org/apache/beam/sdk/transforms/ParDoTest.java:1526]: MAX_ATTEMPTS * timeout >= 300 seconds. I don't believe it should flake but if it does I'll have to figure out a different way to design this test.

File: sdks/java/core/src/test/java/org/apache/beam/sdk/transforms/SplittableDoFnTest.java (hunk 8 above)
Comment (lukecwik) [reply sdks/java/core/src/test/java/org/apache/beam/sdk/transforms/SplittableDoFnTest.java:861]: MAX_ATTEMPTS * timeout >= 300 seconds. I don't believe it should flake but if it does I'll have to figure out a different way to design this test.
